### ISSUE TITLE
Slicing on index creation should be configurable

### ISSUE THREAD

**Issue by malamut2** [BaseXdb/basex#647]
Whenever BaseX is building an index, it is evaluating the memory using Runtime.totalMemory(), Runtime.freeMemory(), and Runtime.maxMemory(). Whenever a fixed threshold is crossed, index information is flushed to disk, System.gc() is called (possibly several times), and index building continues. In the end, all flushed index information is combined into the final index. If the calls to System.gc() do not yield a sufficiently large amount of free memory, OutOfMemory is thrown.

This behaviour is undesirable for a number of reasons, among them:
- Depending on actual GC implementation, results may vary. For example, we have experienced reasonable configurations where the JVM itself threw OOM before the index builder had decided to flush anything to disk.
- In a multithreading environment, the index builder thread may run into trouble due to memory used not by itself, but by some other thread (so that flushing to disk does not free a reasonable amount of memory). That other thread may be, for example, another index builder for another database, or a large query operating on anther database. In the past, with global locking, this has only been an issue for people like me who use more than one call to new Context(), or who embed the BaseX library into a larger Java application. But with the recently introduced database-level locking, this may well become an issue to everyone.
- Calls to System.gc() are generally undesirable because they induce Stop-the-World garbage collections which may take a considerable amount of time, easily a minute or more on large memory systems. This is unsuitable for any sort of server which is supposed to deliver its data on a real-time basis. In fact, some people use the VM option -XX:-DisableExplicitGC to disable well-meant but harmful library calls to System.gc().
- OutOfMemory should never be thrown unless system memory is really exhausted. Libraries should never throw such exceptions on their own, only the JVM itself should throw them. OutOfMemory is usually a good reason to terminate the application without asking further questions, which is why the VM option -XX:OnOutOfMemoryError exists.
- FTIndex seems to rely on its data being flushed to disk before it grows too large. When offering our application a large amount of memory, we ran into the following exceptions (based on line number from a BaseX 7.3 snapshot):

``` java
java.io.IOException: Negative seek offset
  java.io.RandomAccessFile.seek(Native Method)
  org.basex.io.random.DataAccess.cursor(DataAccess.java:229)
  org.basex.io.random.DataAccess.readBytes(DataAccess.java:188)
  org.basex.index.ft.FTList.token(FTList.java:126)
  org.basex.index.ft.FTList.next(FTList.java:85)
  org.basex.index.ft.FTBuilder.merge(FTBuilder.java:259)
  org.basex.index.ft.FTBuilder.write(FTBuilder.java:161)
  org.basex.index.ft.FTBuilder.index(FTBuilder.java:94)
  org.basex.index.ft.FTBuilder.build(FTBuilder.java:102)
  org.basex.index.ft.FTBuilder.build(FTBuilder.java:1)
  org.basex.core.cmd.ACreate.create(ACreate.java:122)
  org.basex.core.cmd.CreateIndex.run(CreateIndex.java:60)
  org.basex.core.Command.run(Command.java:363)
  org.basex.core.Command.exec(Command.java:342)
  org.basex.core.Command.execute(Command.java:78)
  org.basex.core.Command.execute(Command.java:90)
  ...
```

and soon after:

``` java
java.lang.OutOfMemoryError: Java heap space
  org.basex.index.ft.FTList.next(FTList.java:92)
  org.basex.index.ft.FTBuilder.merge(FTBuilder.java:259)
  org.basex.index.ft.FTBuilder.write(FTBuilder.java:161)
  org.basex.index.ft.FTBuilder.index(FTBuilder.java:94)
  org.basex.index.ft.FTBuilder.build(FTBuilder.java:102)
  org.basex.index.ft.FTBuilder.build(FTBuilder.java:1)
  org.basex.core.cmd.ACreate.create(ACreate.java:122)
  org.basex.core.cmd.CreateIndex.run(CreateIndex.java:60)
  org.basex.core.Command.run(Command.java:363)
  org.basex.core.Command.exec(Command.java:342)
  org.basex.core.Command.execute(Command.java:78)
  org.basex.core.Command.execute(Command.java:90)
  ...
```

This did not happen when using a smaller amount of memory. Obviously, flushing the data in time to disk avooided some integer overflow.

Bottom line: we need to be able to configure fixed amounts of some sort which BaseX will use to decide when to flush index data to disk. The undocumented SINGLEGC parameter helps somewhat but is not sufficient. The current behaviour may be kept as the default until reasonable, easily configurable defaults for the new fixed amounts have been established.

I'd like to tackle this issue myself. Information about the file format and algorithms of the index files, including FT, would be welcome for that. Also, I'd like to know whether the current FT classes are still valid or will soon be replaced.
